File: hbp100/extractors/medical.py
import re
from typing import List, Set

from hbp100.extractors.base import BaseExtractor
from hbp100.schemas.entity import Entity, EntityType


class HospitalExtractor(BaseExtractor):
    def _compile_patterns(self):
        self._patterns = [
            re.compile(r'\b(?:Hospital|Medical Center|Clinic)[:\s]+([A-Z][a-zA-Z\s]+?)(?=\s+[A-Z]|$|[,.]|\n)', re.IGNORECASE),
            re.compile(r'\b([A-Z][a-z]+(?:\s+[A-Z][a-z]+){0,2}\s+(?:Hospital|Medical Center|Clinic))\b'),
        ]

    @property
    def supported_types(self) -> List[EntityType]:
        return [EntityType.HOSPITAL]

    def extract(self, text: str) -> List[Entity]:
        self._validate_text(text)
        entities = []
        detected_values: Set[str] = set()

        for pattern in self._patterns:
            for match in pattern.finditer(text):
                try:
                    value = match.group(1).strip()
                except IndexError:
                    value = match.group(0).strip()

                if value in detected_values or len(value) < 5:
                    continue
                if value.upper() in {"HOSPITAL", "MEDICAL CENTER", "CLINIC"}:
                    continue
                if len(value.split()) > 5:
                    continue

                start, end = match.span(1) if match.groups() else match.span()
                entity = Entity(
                    type=EntityType.HOSPITAL,
                    value=value,
                    start=start,
                    end=end,
                    confidence=0.80,
                )
                entities.append(entity)
                detected_values.add(value)

        return entities

File: src/extractors/medical.rs
use crate::interfaces::EntityExtractor;
use crate::schemas::{Entity, EntityType};
use regex::Regex;
use std::collections::{HashMap, HashSet};

pub struct MedicalExtractor {
    patterns: Vec<Regex>,
}

impl MedicalExtractor {
    pub fn new() -> Self {
        let mut extractor = Self {
            patterns: Vec::new(),
        };
        extractor.compile_patterns();
        extractor
    }
}

impl Default for MedicalExtractor {
    fn default() -> Self {
        Self::new()
    }
}

impl MedicalExtractor {
    fn compile_patterns(&mut self) {
        self.patterns = vec![
            Regex::new(
                r"(?i)\b(?:Hospital|Medical Center|Clinic)[:\s]+([A-Z][a-zA-Z\s]+?)(?:\s+[A-Z]|$|[,.]|\n)"
            ).unwrap(),
            Regex::new(
                r"(?i)\b([A-Z][a-z]+(?:\s+[A-Z][a-z]+){0,2}\s+(?:Hospital|Medical Center|Clinic))\b"
            ).unwrap(),
        ];
    }
}

impl EntityExtractor for MedicalExtractor {
    fn name(&self) -> &str {
        "MedicalExtractor"
    }

    fn supported_types(&self) -> Vec<EntityType> {
        vec![EntityType::Medical]
    }

    fn extract(&self, text: &str) -> Vec<Entity> {
        let mut entities = Vec::new();
        let mut detected = HashSet::new();
        
        for pattern in &self.patterns {
            for caps in pattern.captures_iter(text) {
                let value = if let Some(matched) = caps.get(1) {
                    matched.as_str().trim().to_string()
                } else if let Some(matched) = caps.get(0) {
                    matched.as_str().trim().to_string()
                } else {
                    continue;
                };
                
                if detected.contains(&value) || value.len() < 5 {
                    continue;
                }
                
                let upper = value.to_uppercase();
                if upper == "HOSPITAL" || upper == "MEDICAL CENTER" || upper == "CLINIC" {
                    continue;
                }
                
                let word_count = value.split_whitespace().count();
                if word_count > 5 {
                    continue;
                }
                
                let (start, end) = if let Some(matched) = caps.get(1) {
                    (matched.start(), matched.end())
                } else if let Some(matched) = caps.get(0) {
                    (matched.start(), matched.end())
                } else {
                    continue;
                };
                
                detected.insert(value.clone());
                entities.push(Entity {
                    entity_type: EntityType::Medical,
                    value,
                    start,
                    end,
                    confidence: 0.80,
                    placeholder: None,
                    metadata: HashMap::new(),
                });
            }
        }
        
        entities
    }
}